### ISSUE TITLE
Optimizers in Keras V3 are slower than in V2 when working with sparse gradients

### ISSUE THREAD

**Issue by shkarupa-alex** [keras-team/keras#19992]
Here is a small tests for TF 2.11 vs 2.17 https://colab.research.google.com/drive/1i_rVpDQqCEdNvJ4TOj9t1Pbfmhl7DdtS?usp=sharing

There are 2 optimizer versions in TF 2.11 (keras 2): with tf.raw_ops.XXX and with plain TF. Both versions give same speed when used with jit_compile=True:
239 ms ± 5.31 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
238 ms ± 9.29 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

But in Keras 3 (TF 2.17) Adam (for example) has no any clauses about sparse gradients. So speed result is about 15%-20% worse. 
294 ms ± 6.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

**Comment by mehtamansi29** [comment]
Hi @shkarupa-alex -

I am able to reproduce the issue. Attached [gist](https://colab.sandbox.google.com/gist/mehtamansi29/a4accb6f76f2d2b65b1ae1b52cf6dd0e/19992-optimizers-in-keras-v3-are-slower-than-in-v2-when-working-with-sparse-gradients.ipynb) for your reference. We will look into the issue and update you the same.

**Comment by shkarupa-alex** [comment]
> Hi @shkarupa-alex -
> 
> I am able to reproduce the issue. Speed result decreased but not 15%-20%, speed decreased about 5%-7%. Attached [gist](https://colab.sandbox.google.com/gist/mehtamansi29/a4accb6f76f2d2b65b1ae1b52cf6dd0e/19992-optimizers-in-keras-v3-are-slower-than-in-v2-when-working-with-sparse-gradients.ipynb) for your reference. We will look into the issue and update you the same.

This will depend on variable (embedding matrix) size.
Moreover this makes impact on memory footprint.

**Comment by mattdangerw** [comment]
@hertschuh tagging you for any thoughts on sparse ops!

**Comment by hertschuh** [comment]
@shkarupa-alex ,

First, I want to clarify that Keras 3 does support sparse gradients and sparse updates. And it does it without any "if" clauses in the optimizer code.

The way it works is that the `ops` themselves know how to handle `tf.IndexedSlices`, `tf.SparseTensor`s and regular tensors, and then `assign_add` knows to do a `scatter_add` when it's a `tf.IndexedSlices`.

If you're interested, you can look at the details in https://github.com/keras-team/keras/pull/18692 . The only constraint to support sparse gradients is that you have to use ops (e.g. `ops.add` instead of `+`) and `self.assign_*` instead of `var.assign_*`.

Now, about the performance of this. I'm assuming you were testing on CPU. It is actually difficult to get reliable numbers on CPU, different runs gave me different results. In comparison, GPU and TPU testing gave way more reliable numbers.

|             | TF 2.11 Keras 2 - Legacy Adam | TF 2.11 Keras 2 - Adam | TF 2.17 Keras 3 - Adam |
|:-------|:-------:|:-------:|:-------:|
| **Sparse - CPU** | 315 ms ± 28.5 ms | 376 ms ± 46.6 ms | 383 ms ± 14.9 ms |
| **Dense - CPU** | 1.3 s ± 340 ms | 1.24 s ± 295 ms | 865 ms ± 131 ms |
| **Sparse - GPU** | 225 ms ± 4.49 ms | 239 ms ± 20.5 ms | 10.9 ms ± 3.74 µs |
| **Dense - GPU** | 931 ms ± 17 ms | 1.17 s ± 291 ms | 17.9 ms ± 516 µs |
| **Sparse - TPU** | 148 ms ± 7.66 ms | 148 ms ± 13.6 ms | 230 ms ± 6.09 ms |
| **Dense - TPU** | 234 ms ± 2.45 ms | 222 ms ± 2.17 ms | 219 ms ± 1.8 ms |

GPU was a T4 on colab. TPU was a TPU v2 on Colab.

Keras 3 tends to be faster across the board. It's way faster on GPU, it's always faster with dense gradients. It appears to be a bit slower on CPU and TPU with sparse gradient specifically. I tried a few different optimizations to improve the sparse case, but they all degrade the performance in many other cases.

Now, are you using `Adam` specifically? Or were you using it simply as an example for benchmarking?

**Comment by shkarupa-alex** [comment]
@hertschuh i'm not agree with this statement
> The way it works is that the ops themselves know how to handle tf.IndexedSlices, tf.SparseTensors and regular tensors, and then assign_add knows to do a scatter_add when it's a tf.IndexedSlices.

Here is an example https://colab.research.google.com/drive/1FqHjW1g56EFfsczWYkoBYf2FBLl6-xl3#scrollTo=WMq71DwcuCeR
I've took Adam optimizer code "as is" and add debug printing of tensor types.

As you can see gradient is sparse.
But every op that makes addition/multiplication and etc. of `IndexedSlices` tensor with dense one (`m` or `v` in Adam) results in `Dense` tensor.

If our embedding matrix is very large, that dense tensor (with a lot of zeros) will be large too. And they should require a lot of memory and time to be materialized. Variable assignment should be slow too.

**Comment by hertschuh** [comment]
> @hertschuh i'm not agree with this statement
> 
> > The way it works is that the ops themselves know how to handle tf.IndexedSlices, tf.SparseTensors and regular tensors, and then assign_add knows to do a scatter_add when it's a tf.IndexedSlices.
> 
> Here is an example https://colab.research.google.com/drive/1FqHjW1g56EFfsczWYkoBYf2FBLl6-xl3#scrollTo=WMq71DwcuCeR I've took Adam optimizer code "as is" and add debug printing of tensor types.
> 
> As you can see gradient is sparse. But every op that makes addition/multiplication and etc. of `IndexedSlices` tensor with `Dense` one (like `m` or `v` in Adam) results in `Dense` tensor.

Right. But `m` is dense. And `m` is decremented by `m * (1 - beta1)` and incremented by `gradient * (1 - beta1)`. The first one, `m * (1 - beta1)` is dense, and it will materialize a tensor of the full size of the embedding matrix regardless of whether the gradient is sparse or dense. So there is a full dense update of the variable regardless.

Right now Adam is written like (option A):
```python
self.assign_add(
    m, ops.multiply(ops.subtract(gradient, m), 1 - self.beta_1)
)
```
Meaning:
* one addition between `IndexedSlices` and dense
* one dense update of the variable 

You can rewrite it as (option B):
```python
self.assign_sub(m, ops.multiply(m, 1 - self.beta_1))
self.assign_add(m, ops.multiply(gradient, 1 - self.beta_1))
```
Meaning:
* one `IndexedSlices` update of the variable
* one dense update of the variable 

So in both cases there is a dense update of the variable and an addition between `IndexedSlices` and something dense, the only difference is whether this addition is done **on the variable** or **an intermediary tensor**. Intuitively, it should take about the same amount of time, and it mostly does with sparse gradients. Option B is slightly faster on CPU for sparse gradients, which is the crux of your issue report.

Now, I tried option B everywhere and benchmarked it, and for some reason that I don't understand it is way slower (more than 4x) in the dense gradient case.

Note that the same exact thing applies to `v`

> If our embedding matrix is very large, these dense tensors will be large too. And they should require a lot of memory and time to be materialized. Variable assignment should be slow too.

Right, but this happens regardless, because of the `m * (1 - beta1)`.

> > Now, are you using Adam specifically? Or were you using it simply as an example for benchmarking?
> 
> This is just an example where slots are used.

Whether sparse gradient result in `scatter_add/sub` depends on the math of the optimizer. A lot of optimizers have a dense part that is added to the sparse gradient (like `Adam`). But here you'll see the list and whether they'll do `scatter_add/sub`: https://github.com/keras-team/keras/blob/master/keras/src/optimizers/optimizer_sparse_test.py#L31-L126
For instance `Adagrad` will, so you may prefer that optimizer.

**Comment by shkarupa-alex** [comment]
I'm not agree with you in some points:

1) Variable update should not be `dense`. 
We can take SGD as an simple example: sparse grad == sparse update.

2) "Right, but this happens regardless, because of the m * (1 - beta1)"
We should NOT update variable and slots with a tons of zeros.
We should make sparse math and sparse update

Look at `Adam4` in the end of https://colab.research.google.com/drive/1i_rVpDQqCEdNvJ4TOj9t1Pbfmhl7DdtS?usp=sharing#scrollTo=GGyVOefzgLPy
The math is the same, but we operate just with values that should be updated.

With **10** 000x64 embedding matrix:
- On CPU it is near the twice faster than default version.
- On T4 colab GPU it is near 6 times faster.

With **100** 000x64 embedding matrix on T4 colab GPU i see:
- Default Adam: 84.9 ms ± 43.8 µs per loop
- Sparse-friendly Adam: 1.33 ms ± 90.6 µs

**Comment by hertschuh** [comment]
@shkarupa-alex 

Oh, I see. What you're referring to is what was published as `LazyAdam`:
- https://www.tensorflow.org/addons/tutorials/optimizers_lazyadam
- https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/lazy_adam.py

Quoting the documentation:

> LazyAdam is a variant of the Adam optimizer that handles sparse updates more efficiently. The original Adam algorithm maintains two moving-average accumulators for each trainable variable; the accumulators are updated at every step. This class provides lazier handling of gradient updates for sparse variables. It only updates moving-average accumulators for sparse variable indices that appear in the current batch, rather than updating the accumulators for all indices. Compared with the original Adam optimizer, it can provide large improvements in model training throughput for some applications. However, it provides slightly different semantics than the original Adam algorithm, and may lead to different empirical results.

The emphasis here is that `Adam` *updates accumulators for each trainable variable at every step*. In contrast `LazyAdam` only updates indices in the current batch, so it has *slightly different semantics* and potential *different empirical results*.
So we wouldn't make the regular `Adam` behave this way because it changes the semantics.
